chain/network/src/cache.rs

```diff
@@ -0,0 +1,398 @@
+use chrono::Utc;
+use near_primitives::hash::CryptoHash;
+use near_primitives::network::PeerId;
+use std::collections::btree_map;
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+
+type Time = u64;
+type Size = i64;
+
+/// Cache to store route back messages.
+///
+/// The interface of the cache is similar to a regular HashMap:
+/// elements can be inserted, fetched and removed.
+///
+/// Motivation behind the following (complex) design:
+///
+/// Since this cache is populated with messages that should be routed
+/// back it is important that elements are not removed unless there is
+/// some confidence that the response is not going to arrive.
+///
+/// A naive cache can be easily abused, since a malicious actor can send
+/// void messages that should be routed back, letting users replacing useful
+/// entries on the cache for fake ones, and producing most of the messages
+/// that require route back being dropped.
+///
+/// Solution:
+///
+/// The cache will accept new elements until it is full. If it receives a new
+/// element but it implies going over the capacity, the message to remove is
+/// selected as following:
+///
+/// 1. For every message store the time it arrives.
+/// 2. For every peer store how many message should be routed to it.
+///
+/// First are removed messages that have been in the cache more time than
+/// EVICTED_TIMEOUT. If no message was removed, it is removed the oldest
+/// message from the peer with more messages in the cache.
+///
+/// Rationale:
+///
+/// - Old entries in the cache will be eventually removed (no memory leak).
+/// - If the cache is not at full capacity, all new records will be stored.
+/// - If a peer try to abuse the system, it will be able to allocate at most
+///     $capacity / number_of_active_connections$ entries.
+pub struct RouteBackCache {
+    /// Maximum number of records allowed in the cache.
+    capacity: u64,
+    /// Maximum time allowed before removing a record from the cache.
+    evict_timeout: u64,
+    /// Main map from message hash to time where it was created + target peer
+    /// Size: O(capacity)
+    main: HashMap<CryptoHash, (Time, PeerId)>,
+    /// Number of records allocated by each PeerId.
+    /// The size is stored with negative sign, to order in PeerId in decreasing order.
+    /// Size: O(number of active connections)
+    size_per_target: BTreeSet<(Size, PeerId)>,
```

Comment:
Since size is bounded by `capacity`, why not insert `capacity - size` instead so that we completely avoid negative numbers?